ggml-cuda.cu

```diff
@@ -6082,19 +6093,37 @@ inline void ggml_cuda_op_rope(
     memcpy(&freq_scale, (int32_t *) dst->op_params + 5, sizeof(float));
 
     const float theta_scale = powf(freq_base, -2.0f/n_dims);
-    const float p0 = (((mode & 1) == 0 ? n_past : 0)) * freq_scale;
+    // const float p0 = (((mode & 1) == 0 ? n_past : 0)) * freq_scale;
+
+    GGML_ASSERT(src1->type == GGML_TYPE_I32);
+    GGML_ASSERT(src1->ne[0] == ne2);
+    GGML_ASSERT(src1->backend == GGML_BACKEND_GPU);
+
+    int id;
+    CUDA_CHECK(cudaGetDevice(&id));
+
+    int * pos = nullptr;
+    if ((mode & 1) == 0) {
+        struct ggml_tensor_extra_gpu * src1_extra = (ggml_tensor_extra_gpu *) src1->extra;
+        pos = (int *) src1_extra->data_device[id];
+        if (!src1_extra->copied) {
+            CUDA_CHECK(cudaMemcpyAsync(pos, src1->data, ggml_nbytes(src1), cudaMemcpyHostToDevice, main_stream));
+            src1_extra->copied = true;
+        }
+    }
```

Comment:
With the current codebase I don't think there's much you can do to avoid this. The codebase currently covers constant data being copied to VRAM only before the eval directly from the model file. In all other cases the data is written to VRAM as the output of a tensor. You could of course just not offload `src1` which would cause its data to be copied to VRAM automatically in `ggml_cuda_op_flatten` but that would induce more copies.